gpu_ocean/SWESimulators/FBL.py

```diff
@@ -367,61 +369,39 @@ def __init__(self, \
         """
-        Updates hu according periodic boundary conditions
+        Updates hu according to boundary conditions
         """
 
         # Start with fixing the potential sponge
         self.callSpongeNS(gpu_stream, hu0, 1, 0)
         
-        if (self.boundary_conditions.east == 1 and \
-            self.boundary_conditions.west == 1):
-            if (self.nx_halo > self.nx):
-                print("Closed east-west boundary, but nx_halo > nx")
-                return
+        if self.firstU and (self.boundary_conditions.east == 1 or \
+                            self.boundary_conditions.west == 1 or \
+                            self.boundary_conditions.north == 1 or \
+                            self.boundary_conditions.south == 1):
             
             self.closedBoundaryUKernel.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
                         self.nx, self.ny, \
-                        self.nx_halo, self.ny_halo, \
+                        self.bc_north, self.bc_east, self.bc_south, self.bc_west, \
                         hu0.data.gpudata, hu0.pitch)
 
-        elif (self.boundary_conditions.east == 2):
-            ## Currently, this only works with 0 ghost cells:
-            assert(hu0.nx == hu0.nx_halo), \
-                "The current data does not have zero ghost cells"
-
-            ## Call kernel that swaps the boundaries.
-            if self.firstU:
-                print("Periodic boundary conditions - U")
-                self.firstU = False
-                print("[nx, ny, nx_halo, ny_halo]")
-                print([self.nx, self.ny, self.nx_halo, self.ny_halo])
-                
-            self.periodicBoundaryUKernel.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
+        if (self.boundary_conditions.north == 2):
+            self.periodicBoundaryUKernel_NS.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
                         self.nx, self.ny, \
-                        self.nx_halo, self.ny_halo, \
                         hu0.data.gpudata, hu0.pitch)
-        
-        
-        # If there are ghost cells in north-south direction, update them!
-        # TODO: Generalize to both ghost_north and ghost_south
-        # Updating northern ghost cells
-        if (self.ny_halo > self.ny):
-            if self.firstGhostU:
-                print("Updating U ghosts in north-south")
-                self.firstGhostU = False
-            
-            self.updateGhostCellsUKernel.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
+        if (self.boundary_conditions.east == 2):
+            self.periodicBoundaryUKernel_EW.prepared_async_call(self.global_size, self.local_size, gpu_stream, \
                         self.nx, self.ny, \
-                        self.nx_halo, self.ny_halo, \
                         hu0.data.gpudata, hu0.pitch)
-            
+        
+        self.firstU = False
```

Comment:
Why are these firstU and firstV needed?

Reply:
Earlier:  used for debugging in open boundary conditions. 
Now: closed boundary conditions (setting velocities defined orthogonal to the boundary to zero) is handled within the step kernel. We still call the boundary kernels the prior to the first time step to be sure that this is correct (to avoid user error), but do not call the closed boundary kernels afterwards.